Last time we worked on a classification problem (gender prediction with Logistic Regression and Random Forest Classifier).

 Now we want to shift to a **regression problem** with a simple code example, and also explain why regression is used here, the difference from classification, and the reasoning.

🔹 Difference Between Classification & Regression

* Classification: Predicts categories/labels (e.g., Male or Female, Yes or No).

  Example: Gender classification → Logistic Regression, Random Forest Classifier.

* Regression: Predicts continuous values (e.g., Salary, House Price, Marks).

  Example: Predicting a student’s marks based on study hours → Linear Regression, Random Forest Regressor.

* Reason for change:

  We used classification earlier because the output was categorical (Male/Female).

  Now **we use regression because the output is a continuous numerical value** (not categories).

# **Importing Libraries**

In [67]:
import pandas as pd
import numpy as np
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.ensemble import RandomForestRegressor

Normally, when we work on projects, we download datasets as CSV files from sources like Kaggle or given project datasets.

But in scikit-learn, some very popular **datasets** are already **built-in** so that learners can practice quickly without worrying about downloading files.

Here, we don’t need to import an external dataset, as we are using the one provided by scikit-learn."

# **Loading Dataset**

In [68]:
# Load dataset
housing = fetch_california_housing(as_frame=True)
df = housing.frame
print(df.head())

   MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  MedHouseVal  
0    -122.23        4.526  
1    -122.22        3.585  
2    -122.24        3.521  
3    -122.25        3.413  
4    -122.25        3.422  


In [69]:
df.shape

(20640, 9)

🔹 California Housing Dataset Explained

* Source:

    This dataset is from the 1990 California census.

    Widely used for regression practice in Machine Learning.

* Purpose:

    Predict the median house value (MedHouseVal) in California districts.

    Regression problem → target is a continuous numeric value.

* Number of Samples:

    ~20,640 rows (districts).

    Each row represents a district/block group.

* Number of Features: 8 features (independent variables)

    MedInc → Median income of residents in the district.

    HouseAge → Median age of houses in the district.

    AveRooms → Average number of rooms per household.

    AveBedrms → Average number of bedrooms per household.

    Population → Total population in the district.

    AveOccup → Average number of people per household.

    Latitude → Latitude coordinate of the district.

    Longitude → Longitude coordinate of the district.

* Target Variable:

    MedHouseVal → Median house value (in hundreds of thousands of dollars).

* Data Type:

    All features are numeric (continuous values).

    No categorical variables → no need for encoding.


# **Data Preprocessing**

# Missing values

In [70]:
# Check missing values
print("\nMissing values:\n", df.isnull().sum())


Missing values:
 MedInc         0
HouseAge       0
AveRooms       0
AveBedrms      0
Population     0
AveOccup       0
Latitude       0
Longitude      0
MedHouseVal    0
dtype: int64


Thereare no missing values.

# **Defining Features and Target Variable**

In [71]:
#Features (X) & Target (y)
X = df.drop("MedHouseVal", axis=1)   # independent variables
y = df["MedHouseVal"]               # target: Median House Value

*  X = df.drop("MedHouseVal", axis=1)

    1. Creates the feature set X.

    2. df.drop("MedHouseVal", axis=1) removes the column "MedHouseVal" from the DataFrame.

    3. axis=1 specifies that we are dropping a column, not a row.

    4. X now contains all independent variables (predictors) used for modeling.

* y = df["MedHouseVal"]

    1. Creates the target variable y.

    2. df["MedHouseVal"] selects the column "MedHouseVal" from the DataFrame.

    3. y is the dependent variable (what we want to predict).

Summary:

X → all input features (independent variables).

y → output/target variable (Median House Value).

# **Splitting Data into Training and Testing Sets**

In [72]:
# Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

* Splits the dataset into training and testing sets.

* X → feature set, y → target variable.

* test_size=0.2 means 20% of the data will be used for testing, and 80% for training.

* random_state=42 ensures reproducibility; the split will be the same every time you run it.

* X_train and y_train → used to train the model.

* X_test and y_test → used to evaluate the model on unseen data.

# **Feature Scaling for Regression**

In [73]:
# Feature Scaling (important for regression)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

Scaling ensures all features are on a similar range/scale, which helps regression models converge faster and perform better.

* scaler = StandardScaler()

  1. Creates a StandardScaler object from sklearn.preprocessing.

  2. StandardScaler standardizes features by removing the mean and scaling to unit variance.

  3. Important for regression algorithms (like linear regression, SVR) to perform well.

* X_train_scaled = scaler.fit_transform(X_train)

  1. fit_transform() fits the scaler to X_train (calculates mean & standard deviation) and transforms the training data.

  2. Produces X_train_scaled, a scaled version of X_train.

* X_test_scaled = scaler.transform(X_test)

  1. Uses the same scaling parameters (mean & std from training set) to scale X_test.

  2. Ensures that the test data is transformed consistently with training data.

# **Linear Regression Model**

In [74]:
# Model Training
model = LinearRegression()
model.fit(X_train_scaled, y_train)


LinearRegression()

* model = LinearRegression() → Creates a Linear Regression model to learn the relationship between features and target.

* model.fit(X_train_scaled, y_train) → Trains the model on scaled training data, learning coefficients to predict the target.

In [75]:
# Prediction
y_pred = model.predict(X_test_scaled)

* y_pred = model.predict(X_test_scaled)

   Uses the trained Linear Regression model to make predictions.

   X_test_scaled → scaled test features.

   y_pred → predicted target values for the test set.

# **Linear Regression Model Evaluation**

In [76]:
# Evaluation
print("\nMean Squared Error:", mean_squared_error(y_test, y_pred))
print("R² Score:", r2_score(y_test, y_pred))


Mean Squared Error: 0.5558915986952442
R² Score: 0.575787706032451


* mean_squared_error(y_test, y_pred)

  1. Calculates the Mean Squared Error (MSE) between actual values (y_test) and predicted values (y_pred).

  2. Measures the average squared difference between predicted and true values.

  3. Lower MSE → better model performance.

* r2_score(y_test, y_pred)

  1. Calculates the R² score (coefficient of determination).

  2. Measures how well the model explains the variance in the target variable.

  3. R² ranges from 0 to 1; closer to 1 → better fit.

* Why not accuracy, precision, etc.?

  1. Accuracy, precision, recall are for classification problems (categorical targets).

  2. Here, MedHouseVal is continuous (regression problem), so we use regression metrics like MSE and R².

✅ Summary:

Regression → MSE, R²

Classification → Accuracy, Precision, Recall

* Mean Squared Error (MSE) = 0.556

  1. On average, the squared difference between predicted and actual house values is 0.556.

  2. A higher MSE indicates the **model’s predictions are not very close to actual values**.

* R² Score = 0.576

  1. The model explains only about 57.6% of the variance in the target variable.

  2. This is a **moderate fit, not very strong**.

* Interpretation:

   **Linear Regression is not performing very well** on this dataset.

   To improve performance, we can try a more flexible model like Random Forest Regressor, which can capture non-linear relationships better.

# **Random Forest Model**

In [77]:
# Random Forest Model
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)

* rf_model = RandomForestRegressor(n_estimators=100, random_state=42)

  1. Creates a Random Forest Regressor model.

  2. n_estimators=100 → The forest will consist of 100 decision trees. More trees → usually better performance, but slower training.

  3. random_state=42 → Ensures reproducibility; the results will be the same every time you run the code.

* rf_model.fit(X_train, y_train)

  1. Trains the Random Forest on the training data (X_train, y_train).

  2. The model learns patterns from the data by combining predictions from all 100 trees.

* y_pred_rf = rf_model.predict(X_test)

  1. Uses the trained Random Forest to predict target values for the test set.

  2. y_pred_rf → predicted house values for X_test.

# **Random Forest Model Evaluation**

In [78]:
# Evaluation
print("Random Forest Regressor:")
print("Mean Squared Error:", mean_squared_error(y_test, y_pred_rf))
print("R² Score:", r2_score(y_test, y_pred_rf))

Random Forest Regressor:
Mean Squared Error: 0.2553684927247781
R² Score: 0.8051230593157366


* mean_squared_error(y_test, y_pred_rf)

  1. Calculates the Mean Squared Error (MSE) between actual (y_test) and predicted (y_pred_rf) values.

  2. Measures the average squared difference; lower MSE → better predictions.

* r2_score(y_test, y_pred_rf)

  1. Calculates the R² score, showing how much variance in the target is explained by the model.

  2. R² closer to 1 → better model fit; negative or low values → poor fit.

* Why we use these metrics:

  Regression problem → continuous target, so MSE and R² are appropriate.

  Accuracy, precision, recall → only for classification problems.

* Mean Squared Error (MSE) = 0.255

  1. The average squared difference between predicted and actual values is much lower than Linear Regression (0.556).

  2. Indicates predictions are closer to actual values.

* R² Score = 0.805

  1. The model explains about 80.5% of the variance in the target variable.

  2. Much better than Linear Regression’s R² (0.576), showing a stronger fit.

* **Random Forest performs significantly better than Linear Regression on this dataset**.

* Captures non-linear relationships between features and target, improving accuracy.

* You can now use this model for predictions on new data with more confidence.

# **Making Predictions on New Data (Using Random Forest)**

In [80]:
# New data with column names matching X
new_data = pd.DataFrame([[8.3252, 41, 6.984, 1.023, 322, 2.5556, 37.88, -122.23]],
                        columns=X.columns)

# Make prediction
predicted_value = rf_model.predict(new_data)
print("Predicted Median House Value:", predicted_value[0])

Predicted Median House Value: 4.265793


Your prediction 4.265793 → ~$426,579.

The predicted Median House Value for the new input data is 4.265793.

This is a continuous value, as expected in a regression problem.

* pd.DataFrame([...], columns=X.columns)

   Creates a DataFrame for new input data.

   columns=X.columns ensures the feature names match the training data.

  Avoids warnings when predicting with models trained on DataFrames.

* rf_model.predict(new_data)

  Uses the trained **Random Forest model to predict the target**.

  new_data contains values for all features in the same order as the training set.

  Returns the predicted Median House Value.

* print(predicted_value[0])

  Prints the prediction as a single number, since predict() returns an array.